### PR TITLE
Vanguard Prae Pierce Refactor + Bugfix

### LINKED ISSUE #4612: Praetorian can attack through closed doors
### Testmerges

N/A

### Description of the bug

Praetorians can use the Pierce ability to attack people through a closed door.

### What's the difference with what should have happened?

Xenos shouldn't be able to attack through obstacles that are meant to completely block things.

### How do we reproduce this bug?

1. Be vanguard prae.
2. Use Pierce on someone standing near a closed door, with you on other side of that door.
3. Watch them take damage.

### Issue Bingo

- [X] Issue could be reproduced at least once
- [X] Issue happened in a recent (less than 7 days ago) round
- [X] Couldn't find an existing issue about this (https://github.com/cmss13-devs/cmss13/issues)

### PR DESCRIPTION
**Pull request by Birdtalon**
# About the pull request

Refactors a bit of `praetorian_powers.dm`

Fixes #4612 

Previously the ability would check for `opacity` alone meaning that it could be used through doors with windows in and attack people on the other side. Now checks each turf in turn for any `obj` which has `density` and `throwpass = 0`

Also changed the visible message as the description for this ability says the Praetorian is using it's tail but the visible message says claws.

# Explain why it's good for the game

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>
dreamseeker_2023-10-11_21-26-49
dreamseeker_2023-10-11_21-27-31

</details>

# Changelog
:cl:
refactor: Refactored praetorian pierce ability
fix: Praetorian vanguard can no longer pierce through windowed doors
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm [hunk 1]
```diff
@@ -12,50 +12,52 @@
 	if (!check_and_use_plasma_owner())
 		return
 
+	//X = xeno user, A = target atom
+	var/list/turf/target_turfs = getline2(X, A, include_from_atom = FALSE)
+	var/len = LAZYLEN(target_turfs)
+	if(len > 3)
+		target_turfs = target_turfs.Copy(1, 4)
+
```
Comment (Birdtalon) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:20]: I'd appreciate if someone could provide a more graceful solution here.

I'm trying to draw a line of turfs between the user and their target click but only take the first 3 turfs. `getline2()` doesn't have a limit param. So I just got all the turfs and discarded the ones I don't need. Feels wrong.

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm [hunk 2]
```diff
@@ -12,50 +12,52 @@
...
+				var/obj/object = path_content
+				//If we shouldn't be able to pass through it then stop at this turf
+				if(object.density && !object.throwpass)
+					blocked = TRUE
+					break
+
+				if(istype(object, /obj/structure/window/framed))
+					var/obj/structure/window/framed/W = object
+					if(!W.unslashable)
+						W.deconstruct(disassembled = FALSE)
```
Comment (Birdtalon) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:44]: This check for framed windows was here before so I kept it.

Comment (Cthulhu80) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:44]: https://github.com/cmss13-devs/cmss13/blob/master/.github/guides/STYLES.md#dont-use-abbreviations

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm [hunk 3]
```diff
@@ -12,50 +12,52 @@
...
-		for(var/obj/structure/S in temp)
-			if(istype(S, /obj/structure/window/framed))
-				var/obj/structure/window/framed/W = S
-				if(!W.unslashable)
-					W.deconstruct(disassembled = FALSE)
-
-			if(S.opacity)
-				blocked = TRUE
-				break
+	for(var/turf/path_turf as anything in target_turfs)
```
Comment (Birdtalon) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:25]: I don't know if `as anything` is required here but I based this off of 
https://github.com/cmss13-devs/cmss13/blob/84d29ec1541f13eff680470cd948b5f9ffd9ee56/code/modules/mob/living/carbon/xenomorph/abilities/general_powers.dm#L933-L937

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm [hunk 4]
```diff
@@ -12,50 +12,52 @@
 	if (!check_and_use_plasma_owner())
 		return
 
+	//X = xeno user, A = target atom
+	var/list/turf/target_turfs = getline2(X, A, include_from_atom = FALSE)
```
Comment (Cthulhu80) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:16]: then just rename the vars xeno_user and target_atom, why use abbreviations? I haven't looked at the file, maybe it's plagued with too many abbreviations.

Comment (Birdtalon) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:16]: These vars are defined outside the lines I've actually changed. Again I can change them if needed.

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm (hunk 2 above)
Comment (Birdtalon) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:44]: This is just copypasted code from the original file. I can however change it if needed.

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm (hunk 3 above)
Comment (Cthulhu80) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:25]: for loops have an implicit type check, as anything means it won't type check. If the only thing in path is the turf then i'm fairly certain a type check is not required.

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm (hunk 4 above)
Comment (Cthulhu80) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:16]: I see that, like I said I haven't looked in the file and maybe it's plagued with too many abbreviations and it's kind of outside the scope of this pr. Something hacky you could do, is just var/mob/living/carbon/xenomorph/xeno_user = A or whatever the path is and then the type path for the other var as well, but I must stress, that it's better to do it the right way.

Comment (Cthulhu80) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:16]: Ok, I took a look at the file. You should definitely change the var names for at least that proc that you're working in.

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm (hunk 2 above)
Comment (Cthulhu80) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:44]: yeah, you should fix all the one letter vars in that proc.

Comment (Birdtalon) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm]: Addressed in latest commit.

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm (hunk 4 above)
Comment (Birdtalon) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm]: Renamed these vars in latest commit.

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm [hunk 5]
```diff
@@ -1,74 +1,76 @@
...
-	if(!A || A.layer >= FLY_LAYER || !isturf(X.loc))
+	if(!targetted_atom || targetted_atom.layer >= FLY_LAYER || !isturf(source_xeno.loc))
 		return
 
 	if (!check_and_use_plasma_owner())
 		return
 
+	//X = xeno user, A = target atom
+	var/list/turf/target_turfs = getline2(source_xeno, targetted_atom, include_from_atom = FALSE)
+	var/len = LAZYLEN(target_turfs)
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:17]: just for sanity, please name this something other than `len`

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm [hunk 6]
```diff
@@ -1,74 +1,76 @@
...
-	X.visible_message(SPAN_XENODANGER("[X] slashes its claws through the area in front of it!"), SPAN_XENODANGER("You slash your claws through the area in front of you!"))
-	X.animation_attack_on(A, 15)
-
-	X.emote("roar")
-
-	// Loop through our turfs, finding any humans there and dealing damage to them
-	for (var/mob/living/carbon/H in target_mobs)
-		if (!isxeno_human(H) || X.can_not_harm(H))
+		//Check for structures such as doors
+		for(var/atom/path_content in path_turf.contents)
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:33]: add `as anything`, nothing but atoms can be in `contents`

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm [hunk 7]
```diff
@@ -1,74 +1,76 @@
...
+			if(istype(path_content, /obj))
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:34]: `isobj(path_content)`

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm [hunk 8]
```diff
@@ -1,74 +1,76 @@
...
+					blocked = TRUE
+					break
+
+				if(istype(object, /obj/structure/window/framed))
+					var/obj/structure/window/framed/framed_window = object
+					if(!framed_window.unslashable)
+						framed_window.deconstruct(disassembled = FALSE)
+
+			//Check for mobs and add them to our target list for damage
+			if(istype(path_content, /mob/living/carbon))
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm:47]: `iscarbon(path_content)`

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm (hunk 1 above)
Comment (Zonespace27) [reply code/modules/mob/living/carbon/xenomorph/abilities/praetorian/praetorian_powers.dm]: This'd involve making a new proc that acts like `getline2()` but takes a numerical param and stops if the amt of turfs is >= the param

**Comment by Birdtalon** [comment]
Requested changes made

**Review by Zonespace27 — APPROVED** [review]
Sorry for not getting to this sooner!

**Pull request merged**